[Fix]  - Adding Light clients page (#585)
polkadot-developers/polkadot-docs: develop/toolkit/parachains/light-clients.md

---
title: Light Clients
description: Learn about Polkadot light clients, their importance, usage, and how they enable robust, cohesive, and secure application composition.
---

# Light Clients

> ## Introduction

Reviewer: This introduction is quite long. I'd scale it back to briefly introduce the topic, what problems it solves, what the guide covers using the usual recipe for intros

Light clients provide a secure, trust-minimized, and efficient method of interacting with a blockchain. Alongside the commonly used JSON Remote Procedure Calls (RPC), they are one of the primary mechanisms for blockchain communication. Polkadot offers first-class support for light clients, which play a critical role in the development of decentralized applications.

> Light clients are applications that fetch the required data they need from a Polkadot node with an associated cryptographic proof to validate the data. This makes it possible to interact with the Polkadot network without running a full node or having to trust remote peers (for further references, check the [Polkadot Specification, Light Client Messages](https://spec.polkadot.network/sect-lightclient#sect-light-msg){target=\_blank} page).

Reviewer 2: This is kinda confusing. We should use a different word:
```
Light clients are `applications`
```

Author: I rephrased the intro a bit, so this should be fixed by [35d5abf](https://github.com/polkadot-developers/polkadot-docs/pull/585/commits/35d5abf613bb64d5bd4bb4f8b1105b5064bc723c)

```mermaid
flowchart LR
DAPP([dApp])-- Query Account Info -->LC([Light Client])
LC -- Request --> FN(((Full Node)))
LC -- Response --> DAPP
FN -- Response (validated via Merkle proof) --> LC
```

In the diagram above, the decentralized application queries on-chain account information through the light client. The light client, which runs as part of the application, requires minimal memory and computational resources. It uses Merkle proofs to verify the state retrieved from a full node in a trust-minimized manner. Polkadot-compatible light clients utilize [warp syncing](https://spec.polkadot.network/sect-lightclient#sect-sync-warp-lightclient){target=\_blank}, which downloads only block headers.

Light clients can quickly verify the blockchain's state, including [GRANDPA finality](/polkadot-protocol/glossary#grandpa){target=\_blank} justifications.

!!!info "What does it mean to be 'trust-minimized'?"
    "Trust-minimized" means that the light client does not need to fully trust the full node from which it retrieves state. This is achieved through the use of Merkle proofs, which allow the light client to verify the correctness of the state by checking the Merkle tree root.

> ## JSON RPC and Light Client Comparison

Reviewer: We use JSON-RPC with a hyphen. Please update all instances in this guide

Another common method of communication between a user interface (UI) and a node is through the JSON RPC protocol. Generally, the UI retrieves information from the node, fetches network or [pallet](/polkadot-protocol/glossary#pallet){target=\_blank} data, and interacts with the blockchain. This is typically done in one of two ways:

1. **User-Controlled Nodes** - the UI connects to a node client installed on the user's machine
    - These nodes are secure, but installation and maintenance can be inconvenient
2. **Publicly Accessible Nodes** - the UI connects to a third-party-owned publicly accessible node client

>     - These nodes are convenient but centralized and less secure. Applications must maintain a list of backup nodes in case the primary node becomes unavailable

Reviewer: Are these items supposed to be in sequential order? If not, please switch to unordered list

---
title: Light Clients

> description: Learn about Polkadot light clients, their importance, usage, and how they enable robust, cohesive, and secure application composition.

Reviewer: ```suggestion
description:Light clients enable secure and efficient blockchain interaction without running a full node. Learn everything you need to know about light clients on Polkadot.
```

---

# Light Clients

## Introduction

Light clients enable secure and efficient blockchain interaction without running a full node. They provide a trust-minimized alternative to JSON-RPC by verifying data through cryptographic proofs rather than blindly trusting remote nodes.

This guide covers:

- What light clients are and how they work
- Their advantages compared to full nodes and JSON-RPC
- Available implementations in the Polkadot ecosystem
- How to use light clients in your applications

Light clients are particularly valuable for resource-constrained environments and applications requiring secure, decentralized blockchain access without the overhead of maintaining full nodes.

> !!!info "Light node or light client?"

Reviewer 3: Great use of admonitions 💯 I think we are introducing a new term. Instead of double quotes, maybe we can use italics instead. wdyt?

> !!!info "Light node or light client?"

Reviewer: ```suggestion
!!!note "Light node or light client?"
```

    The terms "light node" and "light client" are interchangeable. Both refer to a blockchain client that syncs without downloading the entire blockchain state. All nodes in a blockchain network are fundamentally clients, engaging in peer-to-peer communication.

## Light Clients Workflow

> Unlike JSON-RPC interfaces, where an application must either maintain a list of providers or rely on a single node, light clients are not limited to nor dependent on a single node. They use cryptographic proofs to verify the blockchain's state, ensuring it is both up-to-date and accurate. By verifying only block headers, light clients avoid syncing the entire state, making them ideal for resource-constrained environments.

Reviewer 3: ```suggestion
Unlike JSON-RPC interfaces, where an application must maintain a list of providers or rely on a single node, light clients are not limited to or dependent on a single node. They use cryptographic proofs to verify the blockchain's state, ensuring it is up-to-date and accurate. By verifying only block headers, light clients avoid syncing the entire state, making them ideal for resource-constrained environments.
```

```mermaid
flowchart LR
DAPP([dApp])-- Query Account Info -->LC([Light Client])
LC -- Request --> FN(((Full Node)))
LC -- Response --> DAPP
FN -- Response (validated via Merkle proof) --> LC
```

> In the diagram above, the decentralized application queries on-chain account information through the light client. The light client, which runs as part of the application, requires minimal memory and computational resources. It uses Merkle proofs to verify the state retrieved from a full node in a trust-minimized manner. Polkadot-compatible light clients utilize [warp syncing](https://spec.polkadot.network/sect-lightclient#sect-sync-warp-lightclient){target=\_blank}, which downloads only block headers.

Reviewer 3: ```suggestion
In the diagram above, the decentralized application queries on-chain account information through the light client. The light client runs as part of the application and requires minimal memory and computational resources. It uses Merkle proofs to verify the state retrieved from a full node in a trust-minimized manner. Polkadot-compatible light clients utilize [warp syncing](https://spec.polkadot.network/sect-lightclient#sect-sync-warp-lightclient){target=\_blank}, which downloads only block headers.
```

Light clients can quickly verify the blockchain's state, including [GRANDPA finality](/polkadot-protocol/glossary#grandpa){target=\_blank} justifications.

> !!!info "What does it mean to be 'trust-minimized'?"

Reviewer 3: You are using a single quote in the title and double quotes in the paragraph for the same term, trust-minimized.

Because it's a new term, maybe we should use italics for it?  wdyt?

Reviewer: Yeah, I agree. I would remove the single quotes in the title and on the next line italicize _Trust-minimized_

> !!!info "What does it mean to be trust-minimized?"

Reviewer: ```suggestion
!!!note "What does it mean to be trust-minimized?"
```

> !!!info "What does it mean to be _trust-minimized_?"

Reviewer 3: ```suggestion
!!!info "What does it mean to be trust-minimized?"
```
I think it’s best to remove the italics from the admonition's title to maintain consistency throughout the document.

>     "Trust-minimized" means that the light client does not need to fully trust the full node from which it retrieves state. This is achieved through the use of Merkle proofs, which allow the light client to verify the correctness of the state by checking the Merkle tree root.

Reviewer 3: ```suggestion
    "Trust-minimized" means that the light client does not need to fully trust the full node from which it retrieves the state. This is achieved through the use of Merkle proofs, which allow the light client to verify the correctness of the state by checking the Merkle tree root.
```

## JSON-RPC and Light Client Comparison

Another common method of communication between a user interface (UI) and a node is through the JSON-RPC protocol. Generally, the UI retrieves information from the node, fetches network or [pallet](/polkadot-protocol/glossary#pallet){target=\_blank} data, and interacts with the blockchain. This is typically done in one of two ways:

> - **User-Controlled Nodes** - the UI connects to a node client installed on the user's machine

Reviewer: ```suggestion
- **User-controlled nodes** - the UI connects to a node client installed on the user's machine
```

    - These nodes are secure, but installation and maintenance can be inconvenient

> - **Publicly Accessible Nodes** - the UI connects to a third-party-owned publicly accessible node client

Reviewer: ```suggestion
- **Publicly accessible nodes** - the UI connects to a third-party-owned publicly accessible node client
```

    - These nodes are convenient but centralized and less secure. Applications must maintain a list of backup nodes in case the primary node becomes unavailable

While light clients still communicate with [full nodes](/polkadot-protocol/glossary#full-node), they offer significant advantages for applications requiring a secure alternative to running a full node:

| Full Node                                                                                       | Light Client                                                   |
| :---------------------------------------------------------------------------------------------: | :------------------------------------------------------------: |
| Fully verifies all blocks of the chain                                                          | Verifies only the authenticity of blocks                       |
| Stores previous block data and the chain's storage in a database                                | Does not require a database                                    |

> | Installation, maintenance, and execution are resource-intensive and require technical expertise | No installation; typically included as part of the application |

Reviewer 3: ```suggestion
| Installation, maintenance, and execution are resource-intensive and require technical expertise | No installation is typically included as part of the application |
```

## Using Light Clients

The [`smoldot`](https://github.com/smol-dot/smoldot){target=\_blank} client is the cornerstone of light client implementation for Polkadot SDK-based chains. It provides the primitives needed to build light clients and is also integrated into libraries such as [PAPI](#papi-light-client-support).

### PAPI Light Client Support

> The [Polkadot API (PAPI)](/develop//toolkit/api-libraries/papi){target=\_blank} library natively supports light client configurations, powered by [`smoldot`](https://github.com/smol-dot/smoldot){target=\_blank}. This allows developers to connect to multiple chains simultaneously using a light client.

Reviewer 3: ```suggestion
The [Polkadot API (PAPI)](/develop//toolkit/api-libraries/papi){target=\_blank} library natively supports light client configurations powered by [`smoldot`](https://github.com/smol-dot/smoldot){target=\_blank}. This allows developers to connect to multiple chains simultaneously using a light client.
```
